// tcvui/bridge-h5/src/components/bridgeUtil.js
import bridge from './bridge.js';
import  'whatwg-fetch';
// import './track.js';
//统计代码的引用 和初始化的配置 归属项目自行控制，这边只封装统计提供的方法
// const dfn = '_tc_ntv_util';

let wmodulecache = {
    operationPhoto : {}
},
CACHE_ID = 1;
window.wmodulecache = wmodulecache;

function updateUrlData(url,data){
    let adom = document.createElement('a');
    adom.href = url;
    adom.search += (adom.search === '' ?  data : '&' + data);
    return adom.href;
}
var util = {
    /**
     * 
     * @description 获取数据
     * @param {*} param 配置参数
     * @param {*} url 接口地址
     * @example getData({ body:{id:1} },'url')
     */
    getData(param,url){
        if(!param){
            return;
        }
        let options = { ...{ method:'POST', mode: 'cors', cache: 'default' }, ...param},
            bodys = options.reqbody;

        options.body = {};
        options.body.reqbody = bodys;
        
        options.interfaceid && (options.body.interfaceid = options.interfaceid);
        if(options.method.toLocaleLowerCase() === 'get'){
           url = updateUrlData(url, "data="+ JSON.stringify(options.body))
           delete options.body
        }else{
            options.body = JSON.stringify(options.body)
        }
        return new Promise(function(resolve, reject) {
            fetch(url,options).then(function(response){
                if(response.ok) {
                    return response.json();
                }else{
                    reject();
                }
            }).then(function(data){
                resolve(bridge.buildCBData(data))
            })

        });

    },
    /**
     * 上传图片
     * param => {
        "imgCount": "2",
        "uploadType": "0",
        "projectTag": "guoneiyou"
        }
    */
    uploadPhoto(param){
        CACHE_ID++;
        let _cacheId = CACHE_ID;

        return new Promise(function(resolve, reject){
            operationImg({...param, ...{CBPluginName :_cacheId, CBTagName: ''} });

            wmodulecache.operationPhoto.callback = function(data) {
                try {
                    document.domain = wmodulecache.DOMAINCACHE;
                } catch (e) {}
                if(data.CBPluginName + data.CBTagName === _cacheId+''){
                    removePhotoDom();
                    resolve(bridge.buildCBData({
                        imageList: data.uploadImageList
                    }))
                }
            }
        })
        
    },
    setPage({ pagename = "", referpage = document.referrer} ={} ){
        if(!pagename) throw new Error('can not found the pagename in the param');
        window._tcTraObj && _tcTraObj._tcTrackPage(pagename, referpage);
    },
    setEvent({category = '4in1-h5-event', action = '4in1-h5-event', eventId = "", type = ""} = {}){
        if(!eventId || !type) throw new Error('can not found (eventId,type) in the param');
        window._tcTraObj && _tcTraObj._tcTrackEvent(category, action, eventId, type);
    },
    pops (){

        // require.ensure(['../../../Carousel'], () => {
        //     const Comp = require('../../../Carousel');
            
        // },'Carousel')

      

    }
};

//打开传图页面
function operationImg(param, str) {
    removePhotoDom()
    wmodulecache.DOMAINCACHE = document.domain;
    document.domain = 'ly.com';
    var params = encodeURIComponent(JSON.stringify(param)) + '&CBTagName=' + param.CBTagName + '&CBPluginName=' + param.CBPluginName + '&' + str;
    var odiv = document.createElement('div');
    odiv.id = 'operationPhotoPage';
    odiv.style.cssText="position:fixed;top:0;left:0;right:0;bottom:0;z-index:99999";
    odiv.innerHTML = '<iframe class="operation-img-module" src="//appnew.ly.com/fed/uploadImg/?maxlen=2&callback=wmodulecache.operationPhoto.callback&params=' + params + '" style=" height:100%;width:100%;background:#fff;position:absolute;border:none;outline:none;top:0;left:0;box-sizing:border-box;"></iframe>';
    document.body.appendChild(odiv)
    wmodulecache.IFRAME_DOM = odiv
}
function removePhotoDom(){
    if(wmodulecache.IFRAME_DOM){
        wmodulecache.IFRAME_DOM.parentNode.removeChild(wmodulecache.IFRAME_DOM);
        wmodulecache.IFRAME_DOM = null;
    }
}
 
export default util;
