// tcvui/bridge-h5/src/components/bridgeMap.js
import bridge from './bridge.js';

import location from 'tcvui/util/location/location';

// const dfn = '_tc_ntv_map';


var map = {
      location(param = {}){
          
        return new Promise(function(resolve, reject){
            location({
                time:(param.cacheType  != "2" ? 10000 : false),
                useLocal : (param.cacheType  != "2" )
            }).then(function(data){
                resolve(processLocation(data))
            },function(data){
                resolve(processLocation(data))
            });
        })
      }
};
 

function processLocation(data){
     let _data = {};
    if (data.latitude) {
        _data.lat = data.lat;
        _data.lon = data.lng;
        _data.gps = {
            lat:data.latitude,
            lon:data.longitude
        }
        console.log(data)
        _data.poiName = data.poi[0];
        _data.address = data.formatted_address;
        _data.cityName = data.cityName;
        _data.cityId = data.cityId;
        _data.provinceName = data.provinceName;
        _data.provinceId = data.provinceId;
        _data.countryName = data.country;
        _data.districtName = data.district;
    }

    return {
        CBData: JSON.stringify({
            locationInfo: _data,
            selectedPlaceInfo: _data,
            permanentPlaceInfo: {
            }
        })
    }
            
} 
export default map;
